[common] Support modification of bigint type to string type.

zhuangchong opened the pull request:
<!-- Please specify the module before the PR name: [core] ... or [flink] ... -->

### Purpose

<!-- What is the purpose of the change, or the associated issue -->
close #1273
close #1276 

### Tests

<!-- List UT and IT cases to verify this change -->

SchemaEvolutionTest#testUpdateFieldType
CastExecutorTest#testNumericToString

### API and Format

<!-- Does this change affect API or storage format -->

### Documentation

<!-- Does this change introduce a new feature -->

JingsongLi commented:
Thanks @zhuangchong , looks good!~ Can you create an issue to supports all explicit casting?

zhuangchong commented:
> Thanks @zhuangchong , looks good!~ Can you create an issue to supports all explicit casting?

Already supports all explicit casting.

https://github.com/apache/incubator-paimon/blob/998820f6576ddfa0915000d713a65cae47f0ff18/paimon-core/src/main/java/org/apache/paimon/schema/SchemaManager.java#L285-L292

JingsongLi commented:
> Already supports all explicit casting.

String to int, long?

zhuangchong commented:
> > Already supports all explicit casting.
> 
> String to int, long?

Sorry, I overlooked the string to number type conversion, currently supports number to string and number to number (bigint to int ...)

tsreaper commented on paimon-core/src/main/java/org/apache/paimon/casting/CastExecutors.java:
```diff
@@ -97,6 +97,8 @@ public class CastExecutors {
                                         }
                                 }
                             };
+                        case VARCHAR:
+                            return Object::toString;
```
Return type should be `BinaryString`, not `String`. See the following case branches about `CHAR` and `VARCHAR`.

zhuangchong replied:
done.

tsreaper commented on paimon-core/src/main/java/org/apache/paimon/casting/CastExecutors.java:
```diff
@@ -123,6 +125,8 @@ public class CastExecutors {
                                             (Decimal) value,
                                             decimalType.getPrecision(),
                                             decimalType.getScale());
+                        case VARCHAR:
+                            return Object::toString;
```
Same as above.

zhuangchong replied:
done.

tsreaper commented on paimon-core/src/main/java/org/apache/paimon/schema/SchemaManager.java:
```diff
@@ -283,7 +283,7 @@ public TableSchema commitChanges(List<SchemaChange> changes) throws Exception {
                             update.fieldName(),
                             (field) -> {
                                 checkState(
-                                        DataTypeCasts.supportsImplicitCast(
```
By allowing explicit type casting you will allow casting from `string` to `bigint`, or `timestamp` to `tinyint`, etc., etc. Have you implemented all these in `CastExecutors`?

zhuangchong replied:
This PR is to quickly solve the known BUG of users, support bigint conversion string, I have created a new issue to supports all explicit casting.

tsreaper commented on paimon-core/src/test/java/org/apache/paimon/casting/CastExecutorTest.java:
```diff
@@ -144,6 +144,42 @@ public void testDecimalToNumeric() {
     }
 
+    @Test
+    public void testNumericToString() {
```
Cast result should be `BinaryString`, not normal `String`. See `testStringToString`.

zhuangchong replied:
done.

tsreaper commented on paimon-core/src/test/java/org/apache/paimon/catalog/CatalogTestBase.java:
```diff
@@ -603,11 +603,11 @@ public void testAlterTableUpdateColumnType() throws Exception {
                                         Lists.newArrayList(
                                                 SchemaChange.updateColumnType(
-                                                        "col1", DataTypes.STRING())),
+                                                        "col1", DataTypes.DATE())),
```
Why changing this?

zhuangchong replied:
This test case is Alter table update a column type throws Exception when column data type does not support cast, this PR is modified to support BIGINT to STRING, this case is correct, so I changed it to BIGINT to DATE.

tsreaper replied:
Each PR should be complete by its own. This PR is not for solving bugs, but for introducing a new feature. By introducing this new feature do you really want to create new bugs and just leave them behind? Please fix this.

zhuangchong replied:
New support for type cast: String to Number, Boolean to Number or String, Timestamp to Number or String.

tsreaper commented:
I would also prefer an IT case for `ALTER TABLE ... MODIFY` between all possible types. You mistakenly use `String` and this should be easily discovered by IT cases.

zhuangchong commented:
> I would also prefer an IT case for `ALTER TABLE ... MODIFY` between all possible types. You mistakenly use `String` instead of `BinaryString` and this should be easily discovered by IT cases.

@tsreaper 
Thank you very much for your code review, I have modified it and added a test case.

https://github.com/apache/incubator-paimon/blob/7430f0456f84badaac7deca8d5284b9565dd97e6/paimon-flink/paimon-flink-common/src/test/java/org/apache/paimon/flink/SchemaChangeITCase.java#L210-L226

tsreaper reviewed (commented):
Never squash and force-push your commit unless you have to resolve conflicts by rebasing. If you squash and force-push I can't see what has changed since my last review.

JingsongLi commented on paimon-common/src/main/java/org/apache/paimon/utils/TypeUtils.java:
```diff
@@ -170,6 +170,10 @@ public static boolean toBoolean(BinaryString str) {
         throw new RuntimeException("Cannot parse '" + str + "' as BOOLEAN.");
     }
 
+    public static int toInt(Boolean bool) {
```
`Boolean` -> `boolean`?

zhuangchong replied:
done.

JingsongLi commented on paimon-core/src/main/java/org/apache/paimon/casting/CastExecutors.java:
```diff
@@ -215,19 +245,52 @@ public class CastExecutors {
                                                     % DateTimeUtils.MILLIS_PER_DAY);
                         }
+                    case TINYINT:
+                        return value -> (byte) (((Timestamp) value).getMillisecond());
```
Is `getMillisecond` correct? You can take a look to Flink and Spark.

zhuangchong replied:
done.

zhuangchong commented:
The exception caused the CI exception when downloading the jar, and I restarted the PR multiple times to trigger CI to rerun.

JingsongLi reviewed (changes requested):
Hi @zhuangchong , thanks for the contribution.
Left comments.
You need to add more tests, especially for round-trip testing, we need to cover all round-trip tests.
For example, timestamp cast to string, and string cast to timestamp.

The tests are very very important, please put more effort into testing, cover the tests sufficiently first, and even compare the tests in Flink.

JingsongLi reviewed (commented):
Thanks @zhuangchong for updating! Very Nice!

We can document this too, in https://paimon.apache.org/docs/master/how-to/altering-tables/#changing-column-type
We can document the conversion table.

JingsongLi commented on paimon-core/src/main/java/org/apache/paimon/casting/TimestampToStringCastRule.java:
```diff
@@ -0,0 +1,55 @@
+    @Override
+    public CastExecutor<Timestamp, BinaryString> create(DataType inputType, DataType targetType) {
+        final int precision = DataTypeChecks.getPrecision(inputType);
+        if (inputType.is(DataTypeRoot.TIMESTAMP_WITHOUT_TIME_ZONE)) {
```
What about `TIMESTAMP_WITH_LOCAL_ZONE`?

zhuangchong replied:
done.

JingsongLi replied:
I mean we should support `TIMESTAMP_WITH_LOCAL_ZONE` to string too.

zhuangchong replied:
Thanks. Already supports `TIMESTAMP_WITH_LOCAL_ZONE` type conversion.

zhuangchong commented:
> Thanks @zhuangchong for updating! Very Nice!
> 
> We can document this too, in https://paimon.apache.org/docs/master/how-to/altering-tables/#changing-column-type We can document the conversion table.

I have added type conversion docs.


![image](https://github.com/apache/incubator-paimon/assets/37063904/df7b6562-80ba-4be2-a8ce-b677a7904972)

zhuangchong commented:
Please help to review this PR. Thanks @JingsongLi .

JingsongLi reviewed (approved):
Really thank you @zhuangchong , looks good to me now!

zhuangchong commented:
@JingsongLi Thanks.
